nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java

```diff
@@ -326,7 +326,7 @@ private static Schema nullable(final Schema schema) {
             return Schema.createUnion(unionTypes);
         }
 
-        return Schema.createUnion(Schema.create(Type.NULL), schema);
+        return Schema.createUnion(schema, Schema.create(Type.NULL));
```

Comment:
Changed this and our test schemas per this line in the Avro Specification:

> default: A default value for this field, only used when reading instances that lack the field for schema evolution purposes. The presence of a default value does not make the field optional at encoding time. Permitted values depend on the field's schema type, according to the table below. **Default values for union fields correspond to the first schema in the union**.

Reply:
Does this affect fields for which the default value is `null`? I read in a couple of places (perhaps out of date) that in order to have a default value of `null`, then `null` has to be the first type in the union. Is that no longer the case?

Reply:
A nullable type should always be written as `[ "type_here"], "null" ]`. I'm not sure about a nullable type that also has a default value of null. That could be an edge case in the design of avro itself. What I do know is that Avro 1.9 and newer emphatically reject a default value of a real type if you do "null" first.

Reply:
Using this schema, I got the following validation error:

```
{
	"name": "Test",
	"type": "record",
	"fields": [
		{
			"name": "uuid",
			"type": [
				"string", "null"
			],
			"default": null
		}
	]
}
```
<img width="596" alt="Screen Shot 2022-04-06 at 7 36 18 PM" src="https://user-images.githubusercontent.com/108184/162091400-7de1b271-b7cf-4fa2-b0b7-f93e2b6d31e9.png">

AFAICT, my change is correct because that change was in the context of generating a nullable type, and that's how Avro's docs say to do a nullable type. If someone was able to get by using a bug in Avro's validation, I don't think that's something we should patch around.

Reply:
Reread the specification, and I'm not sure there even is a simple "correct" answer. What I do know is this: the sample schemas we had that worked under Avro 1.8, did not work under 1.10 and 1.11 because they were ordered "null", "type" and had a default value that was not null.

My gut says someone may end up having a bad day because of this dependency upgrade.

For example, I reverted this line locally, and this failed in `testCreateAvroSchemaPrimitiveTypes` in `TestAvroTypeUtil`:

```
fields.add(new RecordField("string", RecordFieldType.STRING.getDataType(), "hola",
                Collections.singleton("greeting")));
```

That string field is written turned into `[ "null", "string" ]` which Avro 1.9+ reject. Avro 1.9+ accept the opposite ordering.

This is the stacktrace; clearly it is an Avro bug that they fixed after 1.8 in their validation:

```
org.apache.avro.AvroTypeException: Invalid default for field string: "hola" not a ["null","string"]

	at org.apache.avro.Schema.validateDefault(Schema.java:1581)
	at org.apache.avro.Schema.access$500(Schema.java:92)
	at org.apache.avro.Schema$Field.<init>(Schema.java:552)
	at org.apache.avro.Schema$Field.<init>(Schema.java:588)
	at org.apache.nifi.avro.AvroTypeUtil.buildAvroField(AvroTypeUtil.java:135)
	at org.apache.nifi.avro.AvroTypeUtil.buildAvroSchema(AvroTypeUtil.java:121)
	at org.apache.nifi.avro.AvroTypeUtil.extractAvroSchema(AvroTypeUtil.java:101)
```

Reply:
I think we are ok upgrading to something that actually enforces the spec, our inference stuff will not generate default values so the non-null type is supposed to come first. If the user has an invalid spec that got by previous versions' validation, it's something we can mention in Migration Guidance as something Avro is checking now.

Reply:
We should change inference to typ,null if it isn't in here already

Reply:
Poked around in the code for JSON and CSV readers and didn't see anything that was even creating obviously nullable fields.